bpf: prevent closing of open BPF maps with reference counting

pillai-ashwin opened the pull request:
This commit implements reference counting to prevent BPF maps from being closed while they are still being accessed by operations like DumpWithCallback, DumpPerCPUWithCallback, and DumpReliablyWithCallback.

Changes:
- Added refCount and closeCond fields to the Map struct to track active operations and signal when all references are released
- Added acquireRef() and releaseRef() helper methods for reference management
- Modified Close() to wait for all active references to be released before closing the map
- Updated DumpWithCallback, DumpPerCPUWithCallback, and DumpReliablyWithCallback to use reference counting
- Added comprehensive tests to verify the fix prevents race conditions

This addresses the race condition where a map could be closed while dump operations were still iterating over it, which could lead to crashes or data corruption.

Fixes: #39287

rgo3 commented:
@tommyp1ckles Could you take a look at this? You proposed a fix via hive on the respective issue, would be nice to know if this is a valid alternative.

tommyp1ckles commented on pkg/bpf/map_linux.go:
```diff
@@ -716,6 +753,12 @@ func (m *Map) DumpWithCallback(cb DumpCallback) error {
 	m.lock.RLock()
 	defer m.lock.RUnlock()
 
+	// Acquire a reference to prevent the map from being closed during iteration.
```
Do we need to do a similar `acquireRef()` inside other funcs like BackLookup?

tommyp1ckles commented:
I think this approach is sound.  

The underlying problems stems from us sharing map descriptors via the global registry which is inherently prone to these issues.  In the future I'd like to move towards managing map lifecycles in via the Hive framework instead of having the global map.

This seems like a good way to solve the issue until that time however.

tommyp1ckles commented on pkg/bpf/map_linux.go:
```diff
@@ -664,6 +672,26 @@ func (m *Map) open() error {
+// releaseRef decrements the reference count and signals waiters if it reaches 0.
+// It must be called with m.lock held (either RLock or Lock).
+func (m *Map) releaseRef() {
+	m.refCount--
```
Just a thought, would we want to provide a way to "drain" this such that a map never has an opportunity to close due to a lot of handles having iterate called (or similar).

pillai-ashwin replied:
Close() potentially waiting forever if operations keep starting.

I've added a `draining` flag. When `Close()` is called, it sets `draining = true`

`acquireRef()` now checks both `m.m == nil` and `draining`

This ensures no new operations can start once Close() begins. So Close() is guaranteed to complete once the currently running operations finish.

Added a test `TestPrivilegedMapDraining()` to verify this behavior.

pillai-ashwin committed:
bpf: prevent closing of open maps with reference counting

Add reference counting to prevent BPF maps from being closed while operations
like DumpReliablyWithCallback are still using them. This fixes the race
condition identified in #39287.

The implementation adds:
- refCount and closeCond fields to track active operations
- acquireRef/releaseRef helpers for reference management
- draining flag so Close() completes even if operations keep starting
- Protection for all long-running operations (Dump*, DeleteAll, ClearAll, BatchLookup)

Short operations like Lookup/Update/Delete already hold the lock and don't
need extra protection.

Fixes: #39287
Signed-off-by: Ashwin Pillai <pillaiashwin96@gmail.com>

pillai-ashwin commented:
/test

tommyp1ckles commented on pkg/bpf/map_linux.go:
```diff
@@ -673,6 +710,20 @@ func (m *Map) Close() error {
+		// Wait for all active references to be released before closing.
+		// Initialize closeCond if not already done.
+		if m.closeCond == nil {
+			m.closeCond = sync.NewCond(&m.lock)
```
Put differently, if this is done in a critical section from the Close()'s m.lock.Lock() and defer m.lock.Unlock() on line 713, then how can acquireRef ever be >0 as by the time we get here all other RLocks will have had to have been released

ti-mo commented:
Can we take a step back and consider if this is really the approach we should take?

I left a comment here: https://github.com/cilium/cilium/issues/39287#issuecomment-3574808487. There are already multiple layers of locking and lifecycle management involved, why are we adding refcounting, of all things? Isn't there just a synchronization bug between Dump*() and Close()?

joestringer commented:
Do I read the discussion right that this is not the right approach? If so, maybe we can close this PR out and follow up in a separate issue or PR.
